### PR TITLE
Yet another implement experiment for underscore anonymous function

### PR DESCRIPTION
**Pull request by clouds56**
The original PR #24990 shows a fancy approach creating anonymous function (at syntax stage or lowering), and could handle lots of useful cases
This PR introduces a new technology parsing `_` at parse stage so that could take advance of parens `()` and distinguish of `in(x, A)` and `x in A`.

The idea comes from [scala](https://stackoverflow.com/questions/28125850/what-are-the-rules-to-govern-underscore-to-define-anonymous-function)

1. If `_` is an argument of a explicit call (say `in(_, A)`, but not `_ in A`), the scope will be carried out to context of the call
    * `f(_+1)` would parsed as `f(x->x+1)` while `f(_)+1` would parse to `x->f(x)+1`
2. if `_` is inside an expression delimited by parenthesis `()`, no matter act as precedence or tuple (but not function call), the scope would be innermost paranthesis
    * `f((_))` would be parsed as `f(x->x)`
    * `2(_+1)` would be parsed as `2*(x->x+1)`
    * `(_, _+1)` would be parsed as `x->(x, y->y+1)` (the same as result of scala)
3. specially `|>` would split the expression to several scope
    * `arr |> map(_+1, _) |> sum(_)` would be parsed as `arr |> (y -> map(x->x+1, y)) |> (x->sum(x))`
    * note this is different from `arr |> y -> map(x->x+1, y) |> x->sum(x)` which is `arr |> (y -> map(x->x+1, y) |> x->sum(x))` since `->` is loose on the right
4. toplevel `=` would be treat specially
   * `a, b = _, _+1` parsed to `a, b = (x->x), (x->x+1)`
6. `_` would never cross block
    * `if a _ else _+1 end` would parsed as `if a x->x else x->x+1 end`
7. other things treats equal, largest possible expression would be used, no matter how many calls it would lowered to
    * `1 - _ - 1` parsed to `x -> 1 -  x - 1`

* additionally since we could have `_` in lvalue so that lvalue `_` would leave as is, but still don't use it in rvalue
    * `_ = 1`, `(_::Int) -> 1`, `(_, f(_)) = 1, 2` would be parsed as is
    * but `_ -> _+1` would be parsed as `_ -> x -> x+1` since the latter is right value
* we disable `_` parsing in macrocall for compatibility.
    * `@match ex begin _ => ex end` would be parsed as is

**TL;DR**: it found scope using parenthesis `()` as you read.

Notes:
* since we have no knowledge we are parsing so that we have to parse it as rvalue and convert to lvalue until seeing things like  `=`, `->`
* shall we treat for generators like `[_+x for x in A]` specially?
* most test case in #24990 could be found in `src/test/underscore-github.scm`
    * more test (for compatibility and consistency) see `src/test/syntax.scm`
    * any additional concern of expression result could reply here and we could add it to test list.
* we might have other (maybe more elegant) technology at parsing to achieve this, but this PR keeps most backward compatibility with current AST and/or code.

### CODE REVIEW

**Comment by clouds56** [comment]
Cheatsheet

| expression | current | original |
| ---- | ---- | ---- |
| `map((_), A)` | `map(x->x, A)` | `x->map(x, A)` |
| `filter(f(_) > a, xs)` | `filter(x->f(_)>a, xs)` | the same [[aplavin]](https://github.com/JuliaLang/julia/pull/24990#issuecomment-1193095687) |
| `filter(f(_) in a, xs)` | `filter(x->f(x) in a, xs)` | `filter((x->f(x)) in a, xs)` |
| `1 + _ + 1` | `x -> 1 + x + 1` | the same [[LilithHafner]](https://github.com/JuliaLang/julia/pull/24990#issuecomment-1205853878) |
| `1 - _ - 1` | `x -> 1 - x - 1` | `(x -> 1 + x) + 1` |
| `f(_, c(b(_ + _) + 1))` | `x -> f(x, c(b((y,z) -> y + z) + 1))` | unknown [[aplavin]](https://github.com/JuliaLang/julia/pull/24990#issuecomment-1206123426) |
| `WeakKeyDict{_,_}` |  `(x,y) -> WeakKeyDict{x,y}` | the same [[knuesel]](https://github.com/JuliaLang/julia/pull/24990#issuecomment-1206335408) |
| `_ isa Integer` | `x -> x isa Integer` | the same |
| `_[i]` | `x -> x[i]` | the same
| `!from(_,m)` | `x -> !from(x,m)` | the same |
| `!_.form_c` | `x -> !x.form_c` | the same |
| `2*_, _%3, _^2` | `x->2*x, x->x%3, x->x^2` | the same |
| `_bool(f) = f(_)::Bool` | `_bool(f) = x->f(x)::Bool` | the same |
| `4 <= _ <= 6` | `x -> 4 <= x <= 6` | the same |
| `!(_ isa LineNumberNode)` | *`!(x -> isa LineNumberNode)`* | `x -> !(x isa LineNumberNode)` |
| `view(A, _...)` | *`view(A, x->x...)`* | `x->view(A, x...)` |
| `_ ∉ dims` | `x -> ∉ dims` | the same |
| `sin.(cos.(_))` | `sin.(x->cos.(x))` | `y -> broadcast(x -> sin(cos(x)), y)` [[stevengj]](https://github.com/JuliaLang/julia/pull/24990#issuecomment-353730741) |
| `[_]` | `x -> [x]` | the same [[innerlee]](https://github.com/JuliaLang/julia/pull/24990#issuecomment-353741737) |
| `(_,)` | `x -> (x,)` | the same |
| `(i for i in _)` | `x -> (i for i in x)` | the same |
| `sum(i for i in _)` | `x -> sum(i for i in x)` | `sum(x -> (i for i in x))` |
| `"$(_)"` | `x -> "$(x)"` | the same |
| `(_...,)` | `(x->x...,)` | `x -> (x...,)` |
| `Array{_}(uninitialized, 0)` | `x -> Array{x}(uninitialized, 0)` | N/A |
| `_'` | `x -> x'` | the same [[innerlee]](https://github.com/JuliaLang/julia/pull/24990#issuecomment-354413704) |
| `_ ? true : false` | `x -> x ? true : false` | implementable |
| `map(_.field + 1, v)` | `map(x -> x.field + 1, v)` | `map((x -> x.field) + 1, v)` [[StefanKarpinski]](https://github.com/JuliaLang/julia/pull/24990#issuecomment-431413599) |
| `map(2_ + 1, v)` | `map(x -> 2x + 1, v)` | `map((x -> 2x) + 1, v)` |
| `y + sqrt(_)` | `x -> y + sqrt(x)` | `y + (x -> sqrt(x))` |
| `map(_.a == 0, ...)` | `map(x -> x.a == 0, ...)` | not worked [[mbauman]](https://github.com/JuliaLang/julia/pull/24990#issuecomment-439548889) |
| `filter(_+2 > y, ...)` | `filter(x -> x+2 > y, ...)` | not worked |
| `filter(_ > y-2, ...)` | `filter(x -> x > y-2, ...)` | the same |
| `findall(2*_ in (4,6,8), ...)` | `findall(x -> 2*x in (4,6,8), ...)` | not worked |
| `_ >_< _` | `(x, y, z) -> x > y < z` | the same |
| `2_ + 1` | `x -> 2x + 1` | the same [[StefanKarpinski]](https://github.com/JuliaLang/julia/pull/24990#issuecomment-442977882) |
| `f(3, 2_ + 1)` | `f(3, x -> 2x + 1)` | proposal of a new method |
| `4f(3, 2_ + 1) - 1` | `4f(3, x -> 2x + 1) - 1` | proposal |
| `map(_[end], A)` | `map(x -> [end], A)` | the same [[mbauman]](https://github.com/JuliaLang/julia/pull/24990#issuecomment-478157855) |
| `map(f(_)[1], A)` | `map(x->f(x[1]), A)` | proposal |
| `map(f(_[1]), A)` | *`map(f(x->x[1]), A)`* | the same |
| `df \|> filter(_.age > 50, _)` | `df \|> y -> filter(x -> x.age > 50, y)` | unknown [[bramtayl ]](https://github.com/JuliaLang/julia/pull/24990#issuecomment-600239227) |
| `I[_]` | `x -> I[x]` | the same [[knuesel]](https://github.com/JuliaLang/julia/pull/24990#issuecomment-1206335408) |
| `startswith(_, entryfile * "_")` | `x -> startswith(x, entryfile * "_")` | the same |
| `_:_` | `(x, y) -> x:y` | the same |
| `_:_:_` | `(x, y, z) -> x:y:z` | the same |
| `_ .+ _` | `(x, y) -> x .+ y` | the same |
| `randn(_, T, _)` | `(x, y) -> randn(x, T, y)` | the same |

the table is not necessay the exhaust list of concerned expression in original PR, nor first memtion by author list above, correct me if I missed something

Notes:
* now it doesn't work with `f(_...)` by design, but is implementable for `...`
* "the same" means the result of original PR is the same as current PR
* "unknown" means the result of original implement doesn't shows in the PR, (though I could guess one)
* "proposal" means the result of original implement doesn't work, a working proposal is suggested by the comment author
* the italic items in column "current" means it failed as intended

**Comment by clouds56** [comment]
Unresolved problems:
- [ ] should we ignore `...` in `_...` to apply unwind? (i.e. parse `f(_...)` to `x->f(x...)` or `f(x->x...)`)
- [ ] how `_` behaves in macros
    - [ ] a) for compatibility we leave all `_` as is, since all `@macro` should treat `_` correctly in current code space, and we could have a new syntax for `_` use in macros
        * option1 `@macro _+1 == @_(_+1)` would parse to `@macro _+1 == x->x+1`
        * option2 `@macro _+1 == $_ + 1` would parse to `@macro _+1 == x->x+1`
        * in both option, almost no code would need modification in existing code
    - [ ] b) for insistence, we parse `_` like what outside macro, if macro would like to accept underscore, use other things like `:_`
        * `@macro $_+1 == _+1` would parse to `@macro _+1 == x->x+1`
        * we could choose `$_` or `:_`
    - [ ] c) maybe something like `use __future__: underscore` could be used for switch behavior
- [ ] how `_` behaves in type system, that is to say shall we parse `Array{_}` things like `Array{T} where T`?
    * this is out of scope for current PR, leave it for future to resolve.
    * since now I don't know why but `Array{_} where _` is a valid expression.

Notice it would take some effort to implement these, so I'd like to hear your feedback.

**Comment by LilithHafner** [comment]
Does this PR break macros that rely on the existing parsing behavior of `_`?

**Comment by clouds56** [comment]
@LilithHafner L, no, it stops translating `_` for macros, so "_ is used in rvalue" error would occur when macros not consume `_`.

**Comment by LilithHafner** [comment]
Consuming `_` before macro expansion breaks MacroTools, for example, a package with 2911 dependants.
```julia
pkg> status
Status `/private/var/folders/hc/fn82kz1j5vl8w7lwd4l079y80000gn/T/jl_4hThOk/Project.toml`
  [1914dd2f] MacroTools v0.5.10

julia> using MacroTools
[ Info: Precompiling MacroTools [1914dd2f-81c6-5fcd-8719-6d5c9610ff09]
ERROR: LoadError: Invalid match clause var"#17#_"->begin
        #= /Users/x/.julia/packages/MacroTools/qijNY/src/utils.jl:256 =#
        var"#17#_" => ex
    end
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] clauses(ex::Expr)
   @ MacroTools ~/.julia/packages/MacroTools/qijNY/src/match/macro.jl:42
 [3] var"@match"(__source__::LineNumberNode, __module__::Module, ex::Any, lines::Any)
   @ MacroTools ~/.julia/packages/MacroTools/qijNY/src/match/macro.jl:55
 [4] include(mod::Module, _path::String)
   @ Base ./Base.jl:434
 [5] include(x::String)
   @ MacroTools ~/.julia/packages/MacroTools/qijNY/src/MacroTools.jl:1
 [6] top-level scope
   @ ~/.julia/packages/MacroTools/qijNY/src/MacroTools.jl:11
 [7] include
   @ ./Base.jl:434 [inlined]
 [8] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt64}}, source::Nothing)
   @ Base ./loading.jl:1638
 [9] top-level scope
   @ stdin:1
in expression starting at /Users/x/.julia/packages/MacroTools/qijNY/src/utils.jl:248
in expression starting at /Users/x/.julia/packages/MacroTools/qijNY/src/utils.jl:247
in expression starting at /Users/x/.julia/packages/MacroTools/qijNY/src/MacroTools.jl:1
in expression starting at stdin:1
ERROR: Failed to precompile MacroTools [1914dd2f-81c6-5fcd-8719-6d5c9610ff09] to /Users/x/.julia/compiled/v1.9/MacroTools/jl_muSUac.
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] compilecache(pkg::Base.PkgId, path::String, internal_stderr::IO, internal_stdout::IO, keep_loaded_modules::Bool)
   @ Base ./loading.jl:1789
 [3] compilecache
   @ ./loading.jl:1733 [inlined]
 [4] _require(pkg::Base.PkgId, env::String)
   @ Base ./loading.jl:1421
 [5] _require_prelocked(uuidkey::Base.PkgId, env::String)
   @ Base ./loading.jl:1281
 [6] macro expansion
   @ ./loading.jl:1261 [inlined]
 [7] macro expansion
   @ ./lock.jl:267 [inlined]
 [8] require(into::Module, mod::Symbol)
   @ Base ./loading.jl:1224
```

File: src/julia-parser.scm [hunk 1]
```diff
@@ -711,20 +773,32 @@
 ;; the principal non-terminals follow, in increasing precedence order
 
 (define (parse-block s (down parse-eq))
+(with-underscore-context #t ; no underscore could cross block
```
Comment (clouds56) [inline src/julia-parser.scm:776]:
```suggestion
(with-underscore-context 'stmt ; no underscore could cross block
```
@LilithHafner points out the use case. otherwise it would break code
```julia
@match ex begin
  _ => ex
end
```

File: src/julia-parser.scm [hunk 2]
```diff
@@ -711,20 +773,32 @@
...
   (parse-Nary s down '(#\newline #\;) 'block
-              (lambda (x) (memq x '(end else elseif catch finally))) #t))
+              (lambda (x) (memq x '(end else elseif catch finally))) #t)))
 
 ;; ";" at the top level produces a sequence of top level expressions
 (define (parse-stmts s)
+(with-underscore-context #t ; no underscore could cross top-level stmts
```
Comment (clouds56) [inline src/julia-parser.scm:782]: (with-underscore-context 'stmt ; no underscore could cross top-level stmts

Comment (clouds56) [inline src/julia-parser.scm:782]:
```suggestion
(with-underscore-context 'stmt ; no underscore could cross top-level stmts
```

**Comment by clouds56** [comment]
@LilithHafner thanks for use cases, fixed.

**Comment by LilithHafner** [comment]
Supporting `_` only outside the scope of all macros is also not okay
```julia
julia> function seconds(x)
           map(_[2], x)
       end
seconds (generic function with 1 method)

julia> seconds([[1,2,3], [4,5,6]])
2-element Vector{Int64}:
 2
 5

julia> @inline function seconds(x)
           map(_[2], x)
       end
ERROR: syntax: all-underscore identifier used as rvalue around REPL[22]:2
Stacktrace:
 [1] top-level scope
   @ REPL[22]:1
```

**Comment by clouds56** [comment]
Now we translate `_` to anonymous function when parsing (from text to AST). Since when AST constructed, much lexical information lost (such as difference between `in(_, A)` and `_ in A`, `f((_))` and `f(_)`, etc.)
While macrocall performs on AST.

So there's 2 option
1. still perform translating during parsing, use another annotation inside macro (like `$_`) to indicate `_` is used for anonymous function:
```julia
@inline function seconds(x)
  map($_[2], x)
end
@match ex begin
  _ => ex
end
```
and the macrocall would see
```julia
@inline function seconds(x)
  map(var"#_#1"->var"#_#1"[2], x)
end
@match ex begin
  _ => ex
end
```
2. Add boundary information in AST, which might be a **breaking changes**, parse `map(_[2], x)` to something like
```julia
begin
  # underscore
  map(begin
    # underscore
    _[2]
  end, _)
end
```
then performs AST transform like macrocalls, then perform `_` translate to
```julia
y -> map(x->x[2], y)
```

**Comment by clouds56** [comment]
ping @LilithHafner, any preference over these 2 options?

**Comment by LilithHafner** [comment]
The choice of symbol really matters for option 1. `_^2 + 1` seems more readable than `$_^2 + 1`. How do we explain to new users why the anonymous function symbol is `$_` rather than `_`?

JuliaParser.jl performs lossless parsing, so perhaps that is what you are looking for with option 2—having rich syntax information available after macro expansion gives us a considerable amount of power. It enables us to do both dwim transformations and create complex syntax and weird bugs.

**Comment by clouds56** [comment]
Generally speaking, there must be a part broken (means breaking changes)
a. macro library (like MacroTools)
b. end user code that uses macro library
c. the parsing workflow (like information in AST)

> How do we explain to new users why the anonymous function symbol is $_ rather than _?

Since we would not like to break b, the existing code occupies the symbol `_`

> JuliaParser.jl performs lossless parsing, so perhaps that is what you are looking for with option 2—having rich syntax information available after macro expansion gives us a considerable amount of power.

I could easily add necessary information in AST (break c), but thus some macro library might not work with the new AST (break a).

## New approach
After this PR, the `_` uses for 2 things in macro
1. as "macro use", which means macro process the `_` and generating code without `_`
2. as "anonymous function use" or "anonymous use" in short, which parser process the `_` and generating anonymous function.

So by definition we need 2 different symbols.


My suggestion is to go with 3 steps
1. lend this PR without supporting `_` as anonymous in macro
```julia
@match ex begin
  _ => map(x->x+1, ex)
end
```
2. deprecate `_` in macro use, let the libraries use `:_`/`@_` instead (for `@_` we would also break macros named `_`)
```julia
@match ex begin
  @_ => map(x->x+1, ex)
end
```
3. and after some time migration, use `_` as anonymous in everywhere.
So the code might be
```julia
@match ex begin
  @_ => map(_+1, ex)
end
```

## Notes
I think `:_` for macro usage is more readable then `_` after anonymous enabled, otherwise one might guess `_` means anonymous or processed by macro library when they read code like this:
```julia
@match ex begin
  _ => map(_+1, ex)
end
```

If we **must** not break anything, the option 1 might be the only way
```julia
@match ex begin
  _ => map($_+1, ex)
end
```

**Comment by DilumAluthge** [comment]
This PR has had no activity for three years+, so I'll close it as stale.

@clouds56 If you're still interested in working on this PR, let me know, and I can re-open it.
